### PR TITLE
feat: tesla ev opentelemetry collection process

### PR DESCRIPTION
**Pull request by dedlockdave**
This pr adds the code for capturing Tesla ev metrics with Open Telemetry

### CODE REVIEW

File: ev-open-telemetry/README.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+After this point a browser window will open up in order for you to login 
+with your Tesla credentials.  Upon successful authentication, 
+you will be redirected to a blank page. you must copy and paste the url in your browser 
+into your terminal which will save the cache file to authenticate for you.
+After you complete this process once, the cache.json file will be able to 
+use the refresh token to keep the authenticated session active.
+
+use `ctr+c` to close the process and run your docker-compose network
+
+**Step 2: Run Docket Network**
```
Comment (sunng87) [inline ev-open-telemetry/README.md:35]:
```suggestion
**Step 2: Run Docker Network**
```

File: ev-open-telemetry/README.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    }
+```
+
+## Future Development
+**To add additional metrics to our collection**
+1. create a new class that sub-classes the `MetricCollector` 
+
+2. Add a property to the `VehicleInstrumentor` that contains a reference to this new collector
+
+3. Update the `MetricFetcher` to
```
Comment (sunng87) [inline ev-open-telemetry/README.md:104]: unfinished

File: ev-open-telemetry/README.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+    class GreptimeDB{
+        <<OTLP Backend>>
+    }
+```
+
+## Future Development
+**To add additional metrics to our collection**
+1. create a new class that sub-classes the `MetricCollector` 
+
```
Comment (sunng87) [inline ev-open-telemetry/README.md:101]: remove empty line between list items

File: ev-open-telemetry/README.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+installed. 
+
+You also must have a valid Tesla Login with a registered vehicle.
+This project uses [TeslaPy](https://github.com/tdorssers/TeslaPy) 
+which leverages the Tesla Owner API to gather the vehicle metrics.
+
+
+To run this demo:
+
+**Step 1: Download the cache.json file which requires you to log in to your Tesla account through your browser**
```
Comment (sunng87) [inline ev-open-telemetry/README.md:18]:
```suggestion
**Step 1: Download the `cache.json` file which requires you to log in to your Tesla account through your browser**
```

File: ev-open-telemetry/README.md [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+git clone https://github.com/GreptimeTeam/demo-scene.git
+cd demo-scene/ev-open-telemetry/ev_observer && poetry install
+TESLA_USER_EMAIL={your_tesla_email} poetry run python main.py
+```
+
+After this point a browser window will open up in order for you to login 
+with your Tesla credentials.  Upon successful authentication, 
+you will be redirected to a blank page. you must copy and paste the url in your browser 
+into your terminal which will save the cache file to authenticate for you.
+After you complete this process once, the cache.json file will be able to
```
Comment (sunng87) [inline ev-open-telemetry/README.md:30]:
```suggestion
After you complete this process once, the `cache.json` file will be able to 
```

File: ev-open-telemetry/ev_observer/ev_observer/__init__.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    headers={
+        # "Authorization": f"Basic {b64_auth}",
+        "x-greptime-db-name": db_name,
+    },
+    timeout=5,
+)
+
+metric_reader = PeriodicExportingMetricReader(exporter, scrape_interval * 1000)
+provider = MeterProvider(metric_readers=[metric_reader])
+metrics.set_meter_provider(provider)
```
Comment (sunng87) [inline ev-open-telemetry/ev_observer/ev_observer/__init__.py:47]: In addition to metrics data, we also need to set some scope attributes like car id.

File: ev-open-telemetry/ev_observer/ev_observer/vehicle.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+class TeslaMetricFetcher(AbstractVehicleDataFetcher):
+    def refresh(self) -> EVMetricData:
+        tesla_user_email = os.getenv("TESLA_USER_EMAIL")
+        if not tesla_user_email:
+            print("TESLA_USER_EMAIL env var is not set", file=sys.stderr)
+            exit(1)
+
+        with teslapy.Tesla(tesla_user_email) as tesla:
+            try:
+                vehicles = tesla.vehicle_list()
```
Comment (sunng87) [inline ev-open-telemetry/ev_observer/ev_observer/vehicle.py:68]: we need to use these vehicle information as scope attribute or metric attribute

**Comment by sunng87** [comment]
@dedlockdave Thank you for this demo. 

Follow our zero-dependency philosophy for demos, I wonder if it's possible to make step 1 an init container in our docker compose cluster. That means you can build a docker image with python, poetry and our script, mark it as [init container](https://docs.docker.com/reference/compose-file/services/#init) and use dependency graph to make sure all other services starts until the init container exit successfully.

And for grafana, we can use their provisioning feature to include a built-in datasource and dashboard. You can check flight demo for reference

**Comment by sunng87** [comment]
This part looks good to me. @dedlockdave are you going to add a grafana dashboard in this PR or in next one?

**Comment by dedlockdave** [comment]
> This part looks good to me. @dedlockdave are you going to add a grafana dashboard in this PR or in next one?

@sunng87 yes I will make those commits on Grafana for Part 2 of this article.  I am also thinking to point the demo to Greptime Cloud for part 3 when we do the deployment on Vercel and leave the part 1 just as docker container.

What do you think?

**Comment by sunng87** [comment]
Sounds good!

**Review by sunng87 — APPROVED** [review]

**Pull request merged**